File: javascript/WorldSugar.js
/**
 * utilitário para fornecer um meio simples de criar o mundo
 * 
 * @param cfg.world
 *            instância de b2World para fazermos as simulações
 * @param cfg.ctx
 *            contexto de desenho
 * @param cfg.scale
 *            zoom. Tipo isso
 * @param cfg.debug
 *            se devemos ativar o modo de debug de desenho
 * @param cfg.running
 *            se a simulação começa parada ou não
 */
function WorldSugar(cfg) {

	// publish config options
	this.cfg = cfg;

	function mkBody(x, y, w, h, isDynamic, isCircle) {
		var fixDef = new b2FixtureDef;
		if (isCircle)
			fixDef.shape = new b2CircleShape(w);
		else {
			fixDef.shape = new b2PolygonShape();
			fixDef.shape.SetAsBox(w, h);
		}
		fixDef.density = 1.0;
		fixDef.friction = 0.1;
		fixDef.restitution = 0.1;
		var bodyDef = new b2BodyDef();
		bodyDef.type = isDynamic ? b2Body.b2_dynamicBody : b2Body.b2_staticBody;
		bodyDef.position.Set(x, y);
		var body = cfg.world.CreateBody(bodyDef);
		body.CreateFixture(fixDef);
		return body;
	}

	this.makeCircle = function(x, y, r, isDynamic) {
		return mkBody(x, y, r, 0, isDynamic, true);
	};

	this.makeBox = function(x, y, w, h, isDynamic) {
		return mkBody(x, y, w, h, isDynamic, false);
	};

	this.debugEnabled = function(b) {
		var debugDraw = null;
		if (b) {
			debugDraw = new b2DebugDraw();
			// setup debug draw
			debugDraw.SetSprite(cfg.ctx);
			debugDraw.SetDrawScale(cfg.scale);
			debugDraw.SetFillAlpha(0.7);
			debugDraw.SetLineThickness(0.9);
			debugDraw.SetFlags(//
			b2DebugDraw.e_shapeBit //
					| b2DebugDraw.e_jointBit // 
					| b2DebugDraw.e_pairBit //
			// | b2DebugDraw.e_centerOfMassBit //
			// | b2DebugDraw.e_aabbBit //
			);
		}
		cfg.world.SetDebugDraw(debugDraw);
	};

	this.running = function(p) {
		cfg.running = p;
	};

	this.step = function() {
		if (cfg.running) {
			cfg.world.Step(1 / 60, 10, 10);
			cfg.world.DrawDebugData();
			cfg.world.ClearForces();
		}
	};

	this.debugEnabled(cfg.debug);
}

File: java/com/sugarbox2d/WorldSugar.java
package com.sugarbox2d;

import org.jbox2d.collision.shapes.CircleShape;
import org.jbox2d.collision.shapes.PolygonShape;
import org.jbox2d.common.Vec2;
import org.jbox2d.dynamics.Body;
import org.jbox2d.dynamics.BodyDef;
import org.jbox2d.dynamics.BodyType;
import org.jbox2d.dynamics.FixtureDef;
import org.jbox2d.dynamics.World;

public class WorldSugar {

	private World b2World;
	private boolean running = true;
	private float scale;

	public WorldSugar(World b2World, boolean debug, float scale) {
		this.b2World = b2World;
		this.scale = scale;
	}

	public WorldSugar(World b2World, float scale) {
		this(b2World, false, 0.1f);
	}

	private Body makeBody(float x, float y, float w, float h, boolean isDynamic,
			boolean isCircle) {
		BodyDef bd = new BodyDef();
		bd.position = new Vec2(x, y);
		bd.type = isDynamic ? BodyType.DYNAMIC : BodyType.STATIC;
		Body body = b2World.createBody(bd);
		FixtureDef fd = new FixtureDef();
		if (isCircle) {
			CircleShape c = new CircleShape();
			fd.shape = c;
			c.m_radius = w;
		} else {
			PolygonShape p = new PolygonShape();
			p.setAsBox(w, h);
			fd.shape = p;
		}
		body.createFixture(fd);
		return body;
	}

	public Body makeBox(float x, float y, float w, float h) {
		return makeBox(x, y, w, h, false);
	}

	public Body makeBox(float x, float y, float w, float h, boolean isDynamic) {
		return makeBody(x, y, w, h, isDynamic, false);
	}

	public Body makeCircle(float x, float y, float r) {
		return makeCircle(x, y, r, true);
	}

	public Body makeCircle(float x, float y, float r, boolean isDynamic) {
		return makeBody(x, y, r, 0, isDynamic, true);
	}

	
	public void running(boolean running) {
		this.running = running;
	}

	public void step() {
		if (running) {
			b2World.step(1.0f / 60.0f, 10, 10);
			b2World.drawDebugData();
			b2World.clearForces();
		}
	}
}
